src/metabase/lib/datascript.cljc

```diff
@@ -0,0 +1,1361 @@
+    (or
+      ;; With a breakout or aggregation: return the breakouts followed by the aggregations.
+      (not-empty (concat (stage-breakouts stage)
+                         (stage-aggregations stage)))
+
+      ;; With no breakouts or aggregations: main source then expressions then joins (in order).
+      ;; TODO: Is that actually the right order?
+      (concat (returned-columns-method (:mbql.stage/source stage))
+              (stage-expressions stage)
+              (->> (:mbql.stage/join stage)
+                   (sort-by :mbql/series)
+                   (mapcat returned-columns-method))))))
+
+(defmethod returned-columns-method :metadata/table [table]
+  (->> table
+       :metadata.column/_source
+       (sort-by :mbql/series)))
+
+(defn returned-columns
+  "Given a query and stage number, return the (ordered) list of the columns it will return.
+
+  - MBQL stages with aggregations return all their breakouts followed by all their aggregations, each in order.
+  - MBQL stages without aggregations return the columns from: main source, expressions, explicit joins in order."
+  ;; TODO: Double-check that ordering.
+  [query-entity stage-number]
+  (returned-columns-method (query-stage query-entity stage-number)))
+
+;; ### Visible columns
+;; This reverses the "column groups" flow from the original API - the primary function [[visible-column-groups]] returns
+;; the groups. [[visible-columns]] calls that and `concat`s the column lists of each group. This is more natural and
+;; avoids expensively inferring the groups from the combined column list.
+(defn- root-column
+  "Recursively follow `:metadata.column/mirror` to find the original column.
+
+  **Note:** This should only be used to get the name, types, etc. for a column. It is not safe to use for comparisons
+  or anything else. Even when \"the same\" column is coming from multiple sources, they should be treated as separate.
+
+  Any clause which transforms a column (eg. a breakout) should be considered a new, fresh column and not have
+  `:metadata.column/mirror`. (Breakouts even without bucketing or binning are new columns, since breakouts are sorted
+  and de-duped via `GROUP BY`.)"
+  [column-entity]
+  (if-let [mirror (:metadata.column/mirror column-entity)]
+    (recur mirror)
+    column-entity))
+
+(defn visible-column-groups
+  "Given a query and stage number, return the (ordered) list of the columns which are \"visible\" to that stage,
+  grouped by their origins.
+
+  That's the following columns, in this order:
+  - Columns returned by the main source.
+  - Columns from expressions on this stage.
+  - Columns returned by each explicit join, in order.
+  - Columns implicitly joinable through any FKs which appear in the above, in the same order.
+
+  MBQL stages with aggregations return all their breakouts followed by all their aggregations, each in order.
+  MBQL stages without aggregations return the columns from: main source, explicit joins in order, expressions."
```

Comment:
This is a different order than what was documented above (main source, expressions, joins).

Reply:
Yeah, good catch.